pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1107 @@
+        XorExpression: _handle_xor_node,
+        ExactlyExpression: _handle_exactly_node,
+        AtMostExpression: _handle_at_most_node,
+        AtLeastExpression: _handle_at_least_node,
+        EXPR.EqualityExpression: _handle_equality_node,
+        NotEqualExpression: _handle_not_equal_node,
+        EXPR.InequalityExpression: _handle_inequality_node,
+        EXPR.RangedExpression: _handle_ranged_inequality_node,
+        BeforeExpression: _handle_before_expression_node,
+        AtExpression: _handle_at_expression_node,
+        AlwaysIn: _handle_always_in_node,
+        _GeneralExpressionData: _handle_named_expression_node,
+        ScalarExpression: _handle_named_expression_node,
+    }
+    _var_handles = {
+        IntervalVarStartTime: _before_interval_var_start_time,
+        IntervalVarEndTime: _before_interval_var_end_time,
+        IntervalVarLength: _before_interval_var_length,
+        IntervalVarPresence: _before_interval_var_presence,
+        ScalarIntervalVar: _before_interval_var,
+        IntervalVarData: _before_interval_var,
+        IndexedIntervalVar: _before_indexed_interval_var,
+        ScalarVar: _before_var,
+        _GeneralVarData: _before_var,
+        IndexedVar: _before_indexed_var,
+        ScalarBooleanVar: _before_boolean_var,
+        _GeneralBooleanVarData: _before_boolean_var,
+        IndexedBooleanVar: _before_indexed_boolean_var,
+        _GeneralExpressionData: _before_named_expression,
+        ScalarExpression: _before_named_expression,
+        IndexedParam: _before_indexed_param,# Because of indirection
+        ScalarParam: _before_param
+    }
+
+    def __init__(self, cpx_model, symbolic_solver_labels=False):
+        self.cpx = cpx_model
+        self.symbolic_solver_labels = symbolic_solver_labels
+        self._process_node = self._process_node_bx
+
+        self.var_map = {}
+        self._named_expressions = {}
+        self.pyomo_to_docplex = ComponentMap()
+
+    def initializeWalker(self, expr):
+        expr, src, src_idx = expr
+        walk, result = self.beforeChild(None, expr, 0)
+        if not walk:
+            return False, result
+        return True, expr
+
+    def beforeChild(self, node, child, child_idx):
+        # Return native types
+        if child.__class__ in EXPR.native_types:
+            return False, (_GENERAL, child)
+
+        if child.__class__ in step_func_expression_types:
+            return _step_function_handles[child.__class__](self, child)
+
+        # Convert Vars Logical vars to docplex equivalents
+        if not child.is_expression_type() or child.is_named_expression_type():
+            return self._var_handles[child.__class__](self, child)
+
+        return True, None
+
+    def exitNode(self, node, data):
+        return self._operator_handles[node.__class__](self, node, *data)
+
+    finalizeResult = None
+
+
+def collect_valid_components(model, active=True, sort=None, valid=set(),
+                             targets=set()):
```

Comment:
I worry a bit about the efficiency of this method.  This checks the ctype of each data object and not just at the container level, plus creating explicit lists of every component data in the model...